internal/controller/apiextensions/composite/overlay.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package composite
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+)
+
+var _ resource.Composite = &overlayComposite{}
+var _ runtime.Unstructured = &overlayComposite{}
+
+type overlayComposite struct {
```

Comment:
I am not sure about this. At a high level:

a. Do we really need Server Side Apply for Composite Controller to XR spec? Does it solve a real problem?
b.  (if ^ is "yes") What we're trying to achieve here looks similar to [UnstructuredExtractor](https://github.com/kubernetes/client-go/blob/5a0a4247921dd9e72d158aaa6c1ee124aba1da80/applyconfigurations/meta/v1/unstructured.go#L38) of client-go. Can we leverage it instead? 

Regarding (a), when I compare latest master with this PR, here is what I am seeing in the managedFields of the composite when I follow the steps [here](https://github.com/turkenh/xp-ssa-experiments/tree/main/issue-4162#reproduction-steps).

<img width="1143" alt="Screenshot 2023-12-26 at 14 30 00" src="https://github.com/crossplane/crossplane/assets/9900707/77805054-2f09-4cfe-9d69-dddabd57b81e">

Obviously, the values for **manager** and **operation** is looking better with this PR, but are there any advantage in terms of functionality that would justify the changes here? Can we replace the updates (e.g. [this](https://github.com/crossplane/crossplane/blob/a7a6049900e26b23fe37551a0f0971214bb7f64b/internal/controller/apiextensions/composite/api.go#L265) one) with a json patch given all we need is to patch a specific field (e.g. compositionRef.name) with the expected field owner, e.g. `apiextensions.crossplane.io/composite` ?

Reply:
> Did you have a chance to test this PR with https://github.com/crossplane/crossplane/issues/4162? When I follow the steps [here](https://github.com/turkenh/xp-ssa-experiments/tree/main/issue-4162#reproduction-steps), I am still observing duplicate keys as before.

Yes, and just tried once again and it still works. Also, [the added e2e test](https://github.com/crossplane/crossplane/pull/5144/files#diff-d236b9f66a48e924036c2e8dc50733e0aa54076e023b8d2ebd690987a73f8867R258) checks exactly that issue for regressions.

> a. Do we really need Server Side Apply for Composite Controller to XR spec? Does it solve a real problem?

IMHO, it does - it clearly divides the field ownership between two controllers and hence also avoid update conflicts, because they update two distinct sets of fields.  On the screenshot the difference is not just the manager/operation pair - the right handside owns less fields.

Status subresource should be updated further with `Update()` calls, because the composite controller is the exclusive owner of it as the whole.

SSA is already used by Function composer, so it should be good to consolidate us here.

`overlayComposite` was introduced in order to minimize the changes in the rest of the system, since the composite `.spec` is modified inside composer, not the controller, and at that point we wanted to catch all changes that were performed by a number of configurators/env fetchers and the selected composer.

> What we're trying to achieve here looks similar to [UnstructuredExtractor](https://github.com/kubernetes/client-go/blob/5a0a4247921dd9e72d158aaa6c1ee124aba1da80/applyconfigurations/meta/v1/unstructured.go#L38) of client-go. Can we leverage it instead?

`UnstructuredExtractor` gives back a partial resource, populated with the fields owned by the specified field manager.  Here, during the first reconcile loop we need to add to the composite fields owned by the controller. Given that the fields are still not persisted at the api server, I do not see how we could use the extractor.